### ISSUE TITLE
Stops playing after 10 or 15 minutes

### ISSUE THREAD

**Issue by RyanTG** [brarcher/baby-sleep-sounds#65]
This might be a phone settings issue and not a problem with the app.

With android 7. When I play this through a Bluetooth speaker, it always shuts off after about 10 or 15 minutes. When this happens there's no indication that the Bluetooth connection has failed, and the taskbar states that the app is playing. If I bring up the app, it shows it as stopped.

I turned off battery optimization for the app. But I'm not sure if any other doze-type phone feature is getting in the way. Maybe I should totally disable Greenify.

I could get a log through android studio if that would help.

**Comment by brarcher** [comment]
Reading about how [Greenify](https://www.howtogeek.com/198135/how-to-optimize-your-android-phones-battery-life-using-greenify/) works, it will:

> stop an app from running, using Android’s built-in “Force Stop” mechanism, but it will also prevent that app from starting up again until you start it.

I wonder if it is stopping the Baby Sleep Sounds app after so many minutes of background playing. The notification is cleared by the app when the playback is stopped by the app; if the app were killed it will not have a chance to remove the notification. 

If you able to get logs when this happens, we can determine if the issue is with the app or with Greenify.

**Comment by RyanTG** [comment]
Cool - I'll give you some logs (I might be a little slow with it, though, because... Baby!). And I'll also try tweaking greenify.

**Comment by RyanTG** [comment]
Hmm, I replicated the issue (3rd time) and then plugged the phone into my computer in order to get a log from android studio, and the problem didn't occur. I'm not sure if AS impacts the test by periodically running some kind of debug wake-up.

So I'll try to log or some other way.

Also, regarding Greenify: Greenify works by specifying which apps should be forced stopped from running in the background. So, if I haven't told it to do so for Baby Sleep Sounds, then Greenify shouldn't factor in here.. Unless there's something I don't know.

**Comment by RyanTG** [comment]
https://gist.github.com/RyanTG/720e54dc606ea617a189b7a6594c3633

Log showing me starting playing a sound (actually, not sure if starting playing is actually on here), putting down my phone for about 15 minutes, app gets killed, and then reopening app after it gets killed. 

I used alogcat for logging. No timestamps!?

Let me know of this is helpful. Even if Greenify isn't the culprit, the problem still could be some other phone setting. I'm using lineageos.

**Comment by brarcher** [comment]
> I used alogcat for logging. No timestamps!?

To get timestamps give the "-v time" or "-v threadtime" flags:
```
$ adb logcat -v time
01-21 02:05:54.658 I/ActivityManager( 1904): Process com.android.calculator2 (pid 27811) has died
01-21 02:05:54.658 D/ActivityManager( 1904): cleanUpApplicationRecord -- 27811
```

Looking at your logs (thanks!) this is probably the issue:
```
I/ActivityManager( 3638): Killing 21323:protect.babysleepsounds/u0a110 (adj 901): excessive wake held 900087 during 900087
```

The [commit](https://android.googlesource.com/platform/frameworks/base/+/35f72be%5E!/) which adds that shows it is doing that if the device is believed to be idle. This means that the app probably needs to do something else to show that it should not be killed. I'll look into that.

**Comment by RyanTG** [comment]
I know that commit doesn't show this, but is there any indication that Bluetooth is also a factor? After making that log, I played a sound without using Bluetooth, but just using my phone's speaker. And it ran successfully for four hours.

I usually never use Bluetooth because it seems so cranky..

**Comment by agrueneberg** [comment]
I'm running into the same problem without a Bluetooth connection or Greenify on Android 7: the app stops playing after some amount of time. I switch back and forth between a disabled (nap) and 1hr (at night) sleep timer, so I initially thought that it doesn't pull out the correct sleep timer value before playing, but it stops after seemingly random amounts of time: sometimes after roughly 15 minutes, sometimes after 30 minutes, but most of the time it works fine.

**Comment by brarcher** [comment]
I went looking through the Android source to see why it may be killing the app. It is holding a wakelock so that Android knows to not sleep. However, my understanding is that because the app is simple (has no background Service, no network traffic, not in the foreground) Android does not think it is doing anything interesting and decides to kill it.

When the app is killed, it is in the foreground (visible) or is it in the background (launched something else)?

Has this issue always happened, or only recently? This issue may have started recently, as I updated the app to build against the latest SDK (27). This may mean that previously its behavior was given a pass, but now background processing is more [strictly controlled](https://developer.android.com/about/versions/oreo/background.html).

I think that the solution is to offload the audio playing to a Service running in the foreground. That way Android will believe that some part is still running in the Foreground even if the Activity is not.

**Comment by brarcher** [comment]
I have some changes which move the audio playback to a foreground service. I saw that after 30 minutes the playback continued, even after moving the activity to the background, navigating to other apps, and turning off the screen.

@RyanTG and @agrueneberg, would either of you be interested in trying this build out to see if the issue is resolved?

[baby-sleep-sounds.apk.zip](https://github.com/brarcher/baby-sleep-sounds/files/1693765/app-release.apk.zip)

**Comment by RyanTG** [comment]
I'll try it out tonight! 

I don't have experience with the previous build, so I can't speak to that.

**Comment by agrueneberg** [comment]
Thanks, I'll try it out over the next few days.

> When the app is killed, it is in the foreground (visible) or is it in the background (launched something else)?

I think the app is usually running in the foreground because I'm trying to sneak out as quietly as possible without fumbling with my phone. I have also disabled battery optimizations after it happened a couple of times, and it didn't improve the situation.

**Comment by RyanTG** [comment]
Last night I ran it twice for about 4.5 hours each time. Ran perfectly! The first time I kept it in the foreground and the second time it was in the background.

I wasn't able to so any debugging - not that it needed it - because... I recently damaged the charging port on my phone and now it only charges (or mounts) when the cable is positioned just right. Last night during the first 4.5 hour sesh the app stopped because the battery drained!
